### ISSUE TITLE
False termination reason "agreement bot discovered cancellation from producer"

### ISSUE THREAD

**Issue by linggao** [open-horizon/anax#127]
{
  "name": "Policy for location merged with location amd64",
  "archived": true,
  "current_agreement_id": "0f5406af533819c7f8f2c825e4dee804fe85e824f8d2b56618436aaa5cb8f00a",
  "consumer_id": "stg-agbot-lon02-01.bluehorizon.hovitos.engineering",
  "agreement_creation_time": 1487754299,
  "agreement_accepted_time": 1487754311,
  "agreement_finalized_time": 1487754406,
  "agreement_data_received_time": 1487754393,
  "agreement_terminated_time": 1487754481,
  "agreement_execution_start_time": 1487754331,
  "terminated_reason": 205,
  "terminated_description": "agreement bot discovered cancellation from producer"
}

Here is timeline for this event:
agbot:
09:04:51 send proposal
09:05:04 got reply, send reply ack
09:05:50 recored agreement on bc
09:06:12 send dataverification to exchange
09:06:34 got bc event (agreement created)
09:06:38 got dataverification ack
09:06:44 cannot find agreement on the bc (governance 150 "discovered terminated agreement")
09:06:44 terminate agreement
09:08:05 got bc event (agreement canceled)

device
09:04:58 received proposal
09:04:59 send reply
09:05:11 received replyack
09:06:23 received dataverification
09:06:45 got bc event (agreement created)
09:08:01 got bc event (agreement canceled)
09:08:01 terminate agreement

scmon 
agreement terminated by consumer

**Comment by linggao** [comment]
arch x86 
cycling test 
snap revision is 74
staging.

**Comment by linggao** [comment]
From booz: from first read it looks like ethereum temporarily forgot about the agreement, it's there and then it's not there. This can happen normally when there are forks on the chain, which is why we wait 3 blocks before we read state from the chain. Egan tells me that we never get forks longer than 3 blocks, which is why we have it set that way. Conforming that this is what happened is VERY complicated.
OR you will have to dig in deeper to see if there really is a problem in our code, maybe there is something you missed in the trace log.

From linggao: That explained the reason well.  I did take a deep look at the agbot code. In govenance.go it sets the AgreementFinalizedTime right after it confirms that the bc record is there. On the other branch, it sees AgreementFinalizedTime is set, then it checks if the bc record is still there.  In our case  there was no  bc record detected and I checked there were no bc writes from the device for this agreement. And the scmon confirmed that the all the bc writes for this agreements were from the consumer.

From booz: yeah, so it seems like it just disappeared which points to a fork on the chain. To prove the theory we would need detailed ethereum logs, and logs would have to be from the nodes acting as miners (which is not the device nor agbot). My advice for now is to let this issue sit and see if it happens again so we can understand how often it happens.

There is also something we could do in the code that MIGHT make a fork less likely to trip us up, but it might be more than you are comfortable attempting. You have seen in the anax device side code that it looks for log entries from the chain by using the event_log code in ethblockchain. I never converted the agbot to do the same thing, the agbot uses the VerifyAgreementRecorded method in protocol.go, which is like calling a getter on the blockchain. If we convert the agbot to read events from the chain the same way anax does, this problem might be less likely to trip us up.
